### PR TITLE
A name and selector MUST NOT be defined together

### LINKED ISSUE #974: A name and selector MUST NOT be defined in the same application reference
As per the [spec](https://github.com/k8s-service-bindings/spec#service-binding) in the context of an application reference, there is a normative statement like this :

> A name and selector **MUST NOT** be defined in the same reference.

The status condition can add the reason for failure: `InvalidApplicationReference`

### PR DESCRIPTION
**Pull request by baijum**
A name and selector MUST NOT be defined in the same application
reference

Fix #974

### CODE REVIEW

**Review by pedjak — CHANGES_REQUESTED** [review]
we should not even allow creating service binding instances that have name and selector defined. 

Validation webhooks are our friend here - the check should be implemented there instead.

File: apis/binding/v1alpha1/servicebinding_webhook.go [hunk 1]
```diff
@@ -104,6 +105,10 @@ func (r *ServiceBinding) ValidateUpdate(old runtime.Object) error {
 	if err != nil {
 		log.Error(err, "Update failed")
 	}
+	if r.Spec.Application.Name != "" && r.Spec.Application.LabelSelector != nil && r.Spec.Application.LabelSelector.MatchLabels != nil {
+		log.Error(err, "name and selector MUST NOT be defined in the application reference")
```
Comment (pedjak) [inline apis/binding/v1alpha1/servicebinding_webhook.go:109]: This check should be performed when service binding is created and updated.

**Review by pedjak — CHANGES_REQUESTED** [review]
the checks must exists for both APIs and acceptance tests should verify if they are active.

File: apis/binding/v1alpha1/servicebinding_webhook.go (hunk 1 above)
Comment (baijum) [reply apis/binding/v1alpha1/servicebinding_webhook.go]: Done. 87bbe9f2d5e073d274aa92c5201c85082defc83c

File: apis/spec/v1alpha2/servicebinding_types.go [hunk 2]
```diff
@@ -33,7 +34,7 @@ type ServiceBindingApplicationReference struct {
 	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
 	Name string `json:"name,omitempty"`
 	// Selector is a query that selects the application or applications to bind the service to
-	Selector metav1.LabelSelector `json:"selector,omitempty"`
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
```
Comment (pedjak) [inline apis/spec/v1alpha2/servicebinding_types.go:37]: what is the motivation for this change?

Comment (baijum) [reply apis/spec/v1alpha2/servicebinding_types.go:37]: This is required to support empty values similar to the non-spec API.

Comment (pedjak) [reply apis/spec/v1alpha2/servicebinding_types.go:37]: not following - spec API already has support for referring an api via name or through application selector. What ServiceBinding resource do you have in mind? Can you please share the example?

Comment (baijum) [reply apis/spec/v1alpha2/servicebinding_types.go:37]: I am trying to follow this :
https://github.com/redhat-developer/service-binding-operator/blob/master/apis/binding/v1alpha1/servicebinding_types.go#L153
We are using a pointer there.

Comment (pedjak) [reply apis/spec/v1alpha2/servicebinding_types.go:37]: With that, we are drifting from spec API, right? We should submit this change to the spec first. Can we implement the verification with that change for now?

File: test/acceptance/features/bindAppToService.feature [hunk 3]
```diff
@@ -331,6 +331,86 @@ Feature: Bind an application to a service
...
+                    selector:
+                      matchLabels:
+                        name: backend-operator
+                services:
+                -   group: stable.example.com
+                    version: v1
+                    kind: Backend
+                    name: backend-demo-empty-app
+            """
+        Then Error message is thrown
```
Comment (baijum) [inline test/acceptance/features/bindAppToService.feature:372]: (Feedback from Cabal)
What is the exact error message? Assert against that specific message.

Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:372]: Check for existing function which does that.

File: test/acceptance/features/bindAppToService.feature [hunk 4]
```diff
@@ -331,6 +331,86 @@ Feature: Bind an application to a service
         Then Error message is thrown
         And Service Binding "binding-request-empty-app" is not persistent in the cluster
 
+    @negative
+    Scenario: Service Binding with name and label selector
```
Comment (baijum) [inline test/acceptance/features/bindAppToService.feature:335]: Expand the scenario title.

Comment (baijum) [reply test/acceptance/features/bindAppToService.feature]: Done

File: apis/spec/v1alpha2/servicebinding_types.go (hunk 2 above)
Comment (baijum) [reply apis/spec/v1alpha2/servicebinding_types.go]: spec updated.

File: test/acceptance/features/bindAppToService.feature (hunk 3 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature]: Removed this and followed similar approach used here:
https://github.com/redhat-developer/service-binding-operator/blob/master/test/acceptance/features/immutableBindings.feature

File: apis/binding/v1alpha1/servicebinding_webhook.go [hunk 5]
```diff
@@ -31,12 +32,17 @@ func (r *ServiceBinding) SetupWebhookWithManager(mgr ctrl.Manager) error {
...
+// +kubebuilder:webhook:path=/validate-binding-operators-coreos-com-v1alpha1-servicebinding,mutating=false,failurePolicy=fail,sideEffects=None,groups=binding.operators.coreos.com,resources=servicebindings,verbs=create;update,versions=v1alpha1,name=vservicebinding.kb.io,admissionReviewVersions={v1beta1}
 
 var _ webhook.Validator = &ServiceBinding{}
 
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type
 func (r *ServiceBinding) ValidateCreate() error {
+	if r.Spec.Application.Name != "" && r.Spec.Application.LabelSelector != nil && r.Spec.Application.LabelSelector.MatchLabels != nil {
+		err := errors.New("name and selector MUST NOT be defined in the application reference")
+		log.Error(err, "Create failed")
+		return err
```
Comment (pedjak) [inline apis/binding/v1alpha1/servicebinding_webhook.go:44]: as mentioned in earlier review round, can we avoid code duplication here and in line 59?

File: apis/spec/v1alpha3/servicebinding_webhook.go [hunk 6]
```diff
@@ -38,12 +39,17 @@ func (r *ServiceBinding) SetupWebhookWithManager(mgr ctrl.Manager) error {
...
 
 // TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
-// +kubebuilder:webhook:path=/validate-servicebinding-io-v1alpha3-servicebinding,mutating=false,failurePolicy=fail,sideEffects=None,groups=servicebinding.io,resources=servicebindings,verbs=update,versions=v1alpha3,name=vspecservicebinding.kb.io,admissionReviewVersions={v1beta1}
+// +kubebuilder:webhook:path=/validate-servicebinding-io-v1alpha3-servicebinding,mutating=false,failurePolicy=fail,sideEffects=None,groups=servicebinding.io,resources=servicebindings,verbs=create;update,versions=v1alpha3,name=vspecservicebinding.kb.io,admissionReviewVersions={v1beta1}
 
 var _ webhook.Validator = &ServiceBinding{}
 
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type
 func (r *ServiceBinding) ValidateCreate() error {
+	if r.Spec.Workload.Name != "" && r.Spec.Workload.Selector != nil && r.Spec.Workload.Selector.MatchLabels != nil {
```
Comment (pedjak) [inline apis/spec/v1alpha3/servicebinding_webhook.go:48]: same comment as above.

File: pkg/reconcile/pipeline/context/spec_binding_impl.go [hunk 7]
```diff
@@ -3,13 +3,14 @@ package context
...
 	"fmt"
+
 	"github.com/redhat-developer/service-binding-operator/apis"
 	"github.com/redhat-developer/service-binding-operator/apis/spec/v1alpha3"
 	"github.com/redhat-developer/service-binding-operator/pkg/client/kubernetes"
 	"github.com/redhat-developer/service-binding-operator/pkg/converter"
 	"github.com/redhat-developer/service-binding-operator/pkg/reconcile/pipeline"
 	"github.com/redhat-developer/service-binding-operator/pkg/reconcile/pipeline/context/service"
-	"k8s.io/api/authentication/v1"
+	v1 "k8s.io/api/authentication/v1"
```
Comment (pedjak) [inline pkg/reconcile/pipeline/context/spec_binding_impl.go:13]: this is no-op change, please revert.

File: pkg/reconcile/pipeline/context/spec_binding_impl_test.go [hunk 8]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 	"encoding/base64"
 	e "errors"
+
```
Comment (pedjak) [inline pkg/reconcile/pipeline/context/spec_binding_impl_test.go:7]: nit: whitespace change, please revert.

File: test/acceptance/features/bindAppToService.feature [hunk 9]
```diff
@@ -331,6 +331,195 @@ Feature: Bind an application to a service
...
+                    resource: deployments
+                    labelSelector:
+                      matchLabels:
+                        name: backend-operator
+                services:
+                -   group: stable.example.com
+                    version: v1
+                    kind: Backend
+                    name: backend-demo-empty-app
+            """
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:371]: the scenario is missing assertion part, i.e. we should assert that error is thrown (specific one) when invalid service binding is applied.

File: test/acceptance/features/bindAppToService.feature [hunk 10]
```diff
@@ -331,6 +331,195 @@ Feature: Bind an application to a service
...
+
+    @negative
+    Scenario: Cannot create Service Binding with name and label selector in the spec API
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:374]: annotation this scenario with `@spec` tag.

File: test/acceptance/features/bindAppToService.feature [hunk 11]
```diff
@@ -331,6 +331,195 @@ Feature: Bind an application to a service
...
+                  apiVersion: apps/v1
+                  kind: Deployment
+                  selector:
+                    matchLabels:
+                      name: backend-operator
+                service:
+                  apiVersion: stable.example.com/v1
+                  kind: Backend
+                  name: backend-demo-empty-app
+            """
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:407]: same comment as with the previous scenario.

File: test/acceptance/features/bindAppToService.feature [hunk 12]
```diff
@@ -331,6 +331,195 @@ Feature: Bind an application to a service
...
+                    resource: deployments
+                    labelSelector:
+                      matchLabels:
+                        name: backend-operator
+                services:
+                -   group: stable.example.com
+                    version: v1
+                    kind: Backend
+                    name: backend-demo-empty-app
+            """
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:467]: same comment as above

File: test/acceptance/features/bindAppToService.feature [hunk 13]
```diff
@@ -331,6 +331,195 @@ Feature: Bind an application to a service
...
+
+    @negative
+    Scenario: Cannot update Service Binding with name and label selector in the spec API
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:470]: add `@spec` tag

File: test/acceptance/features/bindAppToService.feature [hunk 14]
```diff
@@ -331,6 +331,195 @@ Feature: Bind an application to a service
...
+                  apiVersion: apps/v1
+                  kind: Deployment
+                  selector:
+                    matchLabels:
+                      name: backend-operator
+                service:
+                  apiVersion: stable.example.com/v1
+                  kind: Backend
+                  name: backend-demo-empty-app
+            """
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:521]: same comment as above, missing assertion.

**Review by pedjak — CHANGES_REQUESTED** [review]

File: apis/spec/v1alpha3/servicebinding_webhook.go (hunk 6 above)
Comment (baijum) [reply apis/spec/v1alpha3/servicebinding_webhook.go:47]: done

File: test/acceptance/features/bindAppToService.feature (hunk 9 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:371]: done

File: test/acceptance/features/bindAppToService.feature (hunk 12 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:469]: done

File: test/acceptance/features/bindAppToService.feature (hunk 13 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:473]: done

File: test/acceptance/features/bindAppToService.feature (hunk 14 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:524]: done

**Review by pedjak — CHANGES_REQUESTED** [review]
/lgtm

please squash commits

**Comment by baijum** [comment]
> please squash commits

Done
